Feat(eos_cli_config_gen): Add global IP NAT support

kornoa opened the pull request:
## Change Summary

Adding support to render IP NAT global EOS configuration.
- ip nat pool ...
- ip nat kernel buffer size ...
- ip nat synchronization
- ip nat translation ...

## Related Issue(s)

Fixes #2724

## Component(s) name

`arista.avd.eos_cli_config_gen`

## Proposed changes
added `ip_nat`

## How to test
structured config:
````yaml
### IP NAT ###
ip_nat:
  kernel_buffer_size: 64
  pools:
    - name: prefix_16
      prefix_length: 16
      ranges:
        - first_ip: 10.0.0.1
          last_ip: 10.0.255.254
        - first_ip: 10.1.0.0
          last_ip: 10.1.255.255
          first_port: 1024
          last_port: 65535
      utilization_log_threshold: 1
    - name: prefix_32
      prefix_length: 32
      ranges:
        - first_ip: 10.2.0.1
          last_ip: 10.2.0.1
    - name: prefix_24
      prefix_length: 24
      ranges:
        - first_ip: 10.3.0.1
          last_ip: 10.3.0.254
        - first_ip: 10.3.1.0
          last_ip: 10.3.1.255
          first_port: 1024
          last_port: 65535
      utilization_log_threshold: 100
  synchronization:
    description: 'test sync config'
    expiry_interval: 60
    local_interface: Ethernet1
    peer_address: 1.1.1.1
    port_range:
      first_port: 1024
      last_port: 65535
      split_disabled: true
    shutdown: true
  translation:
    address_selection:
      any: true
      hash_field_source_ip: true
    counters: true
    low_mark:
      percentage: 50
      host_percentage: 50
    max_entries:
      limit: 100000
      host_limit: 1000
      ip_limits:
        - ip: 10.0.0.1
          limit: 100
        - ip: 10.0.0.2
          limit: 200
    timeouts:
      - protocol: udp
        timeout: 3600
      - protocol: tcp
        timeout: 7200
````
result in EOS CLI config:
````code
!
ip nat translation address selection any
ip nat translation address selection hash field source-ip
ip nat translation udp-timeout 3600
ip nat translation tcp-timeout 7200
ip nat translation max-entries 100000
ip nat translation low-mark 50
ip nat translation max-entries 1000 host
ip nat translation low-mark 50 host
ip nat translation max-entries 100 10.0.0.1
ip nat translation max-entries 200 10.0.0.2
ip nat kernel buffer size 64
ip nat pool prefix_16 prefix-length 16
   range 10.0.0.1 10.0.255.254
   range 10.1.0.0 10.1.255.255 1024 65535
   utilization threshold 1 action log
ip nat pool prefix_32 prefix-length 32
   range 10.2.0.1 10.2.0.1
ip nat pool prefix_24 prefix-length 24
   range 10.3.0.1 10.3.0.254
   range 10.3.1.0 10.3.1.255 1024 65535
   utilization threshold 100 action log
ip nat synchronization
   description test sync config
   expiry-interval 60
   shutdown
   peer-address 1.1.1.1
   local-interface Ethernet1
   port-range 1024 65535
   port-range split disabled
````

## Checklist

### User Checklist
- N/A

### Repository Checklist

- [x] My code has been rebased from devel before I start
- [x] I have read the [**CONTRIBUTING**](https://avd.sh/en/latest/docs/contribution/overview.html) document.
- [x] My change requires a change to the documentation and documentation have been updated accordingly.
- [x] I have updated [molecule CI](https://github.com/aristanetworks/ansible-avd/tree/devel/ansible_collections/arista/avd/molecule) testing accordingly. (check the box if not applicable)

kornoa committed:
Feat(eos_cli_config_gen): Add ip nat pool support (#2724)

kornoa committed:
Feat(eos_cli_config_gen): Add ip nat kernel buffer size support (#2724)

kornoa committed:
Feat(eos_cli_config_gen): Add ip nat synchronization support (#2724)

kornoa committed:
Feat(eos_cli_config_gen): Add ip nat translation support (#2724)

ClausHolbechArista commented on ansible_collections/arista/avd/molecule/eos_cli_config_gen/documentation/devices/ip-nat.md:
````diff
@@ -0,0 +1,102 @@
+   ip address 10.73.255.122/24
+```
+
+## IP NAT
````
Add blank lines before and after all headings

ClausHolbechArista commented on ansible_collections/arista/avd/molecule/eos_cli_config_gen/documentation/devices/ip-nat.md:
```diff
@@ -0,0 +1,102 @@
+Kernel Buffer Size: 64 MB
+
+### NAT Pools
+- Pool Name: **prefix_16**
```
Try to follow the style of router_bgp.peer_groups for creating sections for each nat pool. Please use a table for the settings instead of these lines above.

ClausHolbechArista commented on ansible_collections/arista/avd/molecule/eos_cli_config_gen/documentation/devices/ip-nat.md:
```diff
@@ -0,0 +1,102 @@
+- Port Range: 1024 - 65535
+- Port Range Split: Disabled
+
+### NAT Translation Settings
```
Please build tables with these settings. Same for NAT syncronization.

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/schemas/schema_fragments/ip_nat.schema.yml:
```diff
@@ -0,0 +1,180 @@
+keys:
+  ip_nat:
+    documentation_options:
+      filename: data_model/IP NAT.md
```
```suggestion
      filename: data_model/IP NAT
```

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/schemas/schema_fragments/ip_nat.schema.yml:
```diff
@@ -0,0 +1,180 @@
+                protocol:
+                  type: str
+                  valid_values: ["tcp", "udp"]
+                timeout:
```
Please add a description like `Timeout in seconds` or whatever unit is applicable here.

kornoa committed:
Doc(eos_cli_config_gen): Fix IP NAT schema doc

kornoa committed:
Fix(eos_cli_config_gen): IP NAT device doc

- Respect blank lines before/after headings
- List NAT pools as headings
- Use tables for settings
- Inlcude EOS device configuration

kornoa committed:
Doc(eos_cli_config_gen): IP NAT schema descriptions

gmuloc commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/schemas/schema_fragments/ip_nat.schema.yml:
```diff
@@ -0,0 +1,182 @@
+      pools:
+        type: list
+        primary_key: name
+        convert_types:
+          - dict
```
you can remove this as this is in our schema as we are migrating from 3.x to 4.x
essentially this would be used to support

```
ip_nat:
  pools:
    POOL_NAME:
      [stuff]
```

by converting it to

```
ip_nat:
  pools:
    - name:POOL_NAME
       [stuff]
```

in the code but we don't need to support this for new models.

(Keep the conversion for int to str elsewhere though)

kornoa replied:
That was not really clear to me. But understood now.

kornoa replied:
Fixed in [8899c99](https://github.com/aristanetworks/ansible-avd/pull/2747/commits/8899c9909d24d56209fec75f70e38110c176d56b)

gmuloc commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/schemas/schema_fragments/ip_nat.schema.yml:
```diff
@@ -0,0 +1,182 @@
+                  - str
+                min: 1
+                max: 99
+              host_percentage:
+                type: int
+                convert_types:
+                  - str
+                min: 1
+                max: 99
```
Maybe for this one because the key can look "surprising" for someone used to the CLI you can add a description stating
```
Used to render 'ip nat translation low-mark <host_percentage> host
```

(Happy to be told no on this one - just wondering if this would be obvious for someone coming to AVD)

kornoa replied:
I can add the comment. No problem.

kornoa replied:
Fixed in [8899c99](https://github.com/aristanetworks/ansible-avd/pull/2747/commits/8899c9909d24d56209fec75f70e38110c176d56b)

gmuloc commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/documentation/ip-nat.j2:
```diff
@@ -0,0 +1,108 @@
+## IP NAT
+{%     if ip_nat.kernel_buffer_size is arista.avd.defined %}
+
+| Settings | Value |
```
I know the doc is not really uniform about this in the rest but it makes more sense to keep Setting singular
```suggestion
| Setting | Value |
```

kornoa replied:
Agreed.

kornoa replied:
Fixed in [ece7ab2](https://github.com/aristanetworks/ansible-avd/pull/2747/commits/ece7ab239d6f3915e0b9829b036a53e774724fce)

gmuloc commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos-intended-config.j2:
```diff
@@ -210,6 +210,8 @@
 {% include 'eos/static-routes.j2' %}
 {# IPv6 Static Route #}
 {% include 'eos/ipv6-static-routes.j2' %}
+{# IP NAT #}
+{% include 'eos/ip-nat.j2' %}
```
So without asking you to change anything at this stage, I just realized looking at my test device that the nat config is not in one single place in the running config which means AVD won't be rendering the config in the correct order (which we "_try_" to do as much as possible).

```
ip nat translation max-entries 100000
ip nat translation low-mark 50
ip nat translation max-entries 1000 host
ip nat translation low-mark 50 host
ip nat translation max-entries 200 10.0.0.2
ip nat translation max-entries 100 10.0.0.1
ip nat kernel buffer size 64
!
hostname Host1
!
spanning-tree mode mstp
!
vrf instance mgmt
!
aaa authorization exec default local
!
interface Ethernet1
!
interface Ethernet2
!
interface Ethernet3
!
interface Ethernet4
!
interface Ethernet5
!
interface Ethernet6
!
interface Ethernet7
!
interface Ethernet8
!
interface Management1
!
ip routing
no ip routing vrf mgmt
!
ip route vrf mgmt 0.0.0.0/0 1.1.1.1
!
ip nat pool prefix_16 prefix-length 16
   range 10.0.0.1 10.0.255.254
   range 10.1.0.0 10.1.255.255 1024 65535
   utilization threshold 1 action log
!
ip nat pool prefix_24 prefix-length 24
   range 10.3.0.1 10.3.0.254
   range 10.3.1.0 10.3.1.255 1024 65535
!
ip nat pool prefix_32 prefix-length 32
   range 10.2.0.1 10.2.0.1
!
ip nat synchronization
   description test sync config
   expiry-interval 60
```

an alternative could be to use two templates for the config (keeping one for the documentation) as it is quite straighforward to identify what goes where.

Thoughts @kornoa / @ClausHolbechArista  ?

kornoa replied:
I also considered splitting the NAT config. It is just the question whether the simplicity in Jinja or the order in the EOS config prevails. I will simply follow your guidelines here.

gmuloc replied:
Hi @kornoa - thanks for the other fixes - I discussed this with Claus and we think we should split the Jinja template for eos config rendering for this one to match closer the expected EOS running config.

kornoa replied:
Copy.

kornoa replied:
Fixed in https://github.com/aristanetworks/ansible-avd/pull/2747/commits/dbb7ed0fbc95f0b52259dcef05d26c31291b581c

gmuloc commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/documentation/ip-nat.j2:
```diff
@@ -0,0 +1,108 @@
+| First IP Address  | Last IP Address | First Port | Last Port |
+| ----------------- | --------------- | ---------- | --------- |
+{%             for range in pool.ranges | arista.avd.default([]) %}
+| {{ range.first_ip }} | {{ range.last_ip }} | {{ range.first_port | arista.avd.default('n.a.') }} | {{ range.last_port | arista.avd.default('n.a.') }} |
```
We tend to use `"-"` for unset values rather than n.a.

```suggestion
| {{ range.first_ip }} | {{ range.last_ip }} | {{ range.first_port | arista.avd.default("-") }} | {{ range.last_port | arista.avd.default("-") }} |
```

kornoa replied:
Copy.

kornoa replied:
Fixed in [ece7ab2](https://github.com/aristanetworks/ansible-avd/pull/2747/commits/ece7ab239d6f3915e0b9829b036a53e774724fce)

gmuloc reviewed (commented):
Thanks @kornoa for the PR and the updates after @ClausHolbechArista comments!

Added a couple of my own - some of them more for discussions than changes really.

kornoa committed:
Fix(eos_cli_config_gen): IP NAT device doc

kornoa committed:
Doc(eos_cli_config_gen): IP NAT schema

- Added descriptions
- Removed obsolete type conversion

kornoa committed:
Doc(eos_cli_config_gen): Fix IP NAT schema

kornoa committed:
Fix(eos_cli_config_gen): IP NAT EOS config order

- Split global IP NAT config into two sections

gmuloc reviewed (approved):
LGTM

carlbuchmann reviewed (approved):
LGTM, thank you for this significant contribution!
